packages/expo/cli/api/rest/client.ts

```diff
@@ -0,0 +1,121 @@
+import { getExpoHomeDirectory } from '@expo/config/build/getUserState';
+import { JSONValue } from '@expo/json-file';
+import fetchInstance from 'node-fetch';
+import path from 'path';
+
+import { EXPO_BETA, EXPO_NO_CACHE } from '../../utils/env';
+import { getExpoApiBaseUrl } from '../endpoint';
+import { getAccessToken, getSessionSecret } from '../user/sessionStorage';
+import { FileSystemCache } from './cache/FileSystemCache';
+import { wrapFetchWithCache } from './cache/wrapFetchWithCache';
+import { FetchLike } from './client.types';
+import { wrapFetchWithPrefixUrl } from './wrapFetchWithPrefixUrl';
+
+export class ApiV2Error extends Error {
+  readonly name = 'ApiV2Error';
+  readonly expoApiV2ErrorCode: string;
+  readonly expoApiV2ErrorDetails?: JSONValue;
+  readonly expoApiV2ErrorServerStack?: string;
+  readonly expoApiV2ErrorMetadata?: object;
+
+  constructor(response: {
+    message: string;
+    code: string;
+    stack?: string;
+    details?: JSONValue;
+    metadata?: object;
+  }) {
+    super(response.message);
+    this.expoApiV2ErrorCode = response.code;
+    this.expoApiV2ErrorDetails = response.details;
+    this.expoApiV2ErrorServerStack = response.stack;
+    this.expoApiV2ErrorMetadata = response.metadata;
+  }
+}
+
+/**
+ * An Expo server error that didn't return the expected error JSON information.
+ * The only 'expected' place for this is in testing, all other cases are bugs with the server.
+ */
+export class UnexpectedServerError extends Error {
+  readonly name = 'UnexpectedServerError';
+}
+
+/**
+ * @returns a `fetch` function that will inject user authentication information and handle errors from the Expo API.
+ */
+export function wrapFetchWithCredentials(fetchFunction: FetchLike): FetchLike {
+  return async function fetchWithCredentials(url, options = {}) {
+    if (Array.isArray(options.headers)) {
+      throw new Error('request headers must be in object form');
+    }
+
+    const resolvedHeaders = options.headers || ({} as any);
+
+    const token = getAccessToken();
+    if (token) {
+      resolvedHeaders.authorization = `Bearer ${token}`;
+    } else {
+      const sessionSecret = getSessionSecret();
+      if (sessionSecret) {
+        resolvedHeaders['expo-session'] = sessionSecret;
+      }
+    }
+
+    const results = await fetchFunction(url, {
+      ...options,
+      headers: resolvedHeaders,
+    });
+
+    if (results.status >= 400 && results.status < 500) {
```

Comment:
So far the needs don't require anything more. I don't want to over optimize and have code that exists for cryptic reasons.

Reply:
It's for correct behavior and making it easier to debug when something unexpected happens. The API v2 client will be more robust if it uses the response content to determine whether an API v2 error occurred rather than using the underlying HTTP status.